controllers/nugus_controller/nugus_controller.cpp

```diff
@@ -55,76 +81,317 @@ class NUgus : public webots::Robot {
+                                    touchSensor->enable(sensor_time_step);
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    void sendData() {
+        // Create the SensorMeasurements message
+        auto sensorMeasurements = std::make_unique<SensorMeasurements>();
 
-                // Send a message to the client
-                msg.set_num(current_num);
+        //   uint32 time = 1;  // time stamp at which the measurements were performed expressed in [ms]
+        // repeated Message messages = 2;
+        // repeated AccelerometerMeasurement accelerometers = 3;
+        // repeated BumperMeasurement bumpers = 4;
+        // repeated CameraMeasurement cameras = 5;
+        // repeated ForceMeasurement forces = 6;
+        // repeated Force3DMeasurement force3ds = 7;
+        // repeated Force6DMeasurement force6ds = 8;
+        // repeated GyroMeasurement gyros = 9;
+        // repeated PositionSensorMeasurement position_sensors = 10;
 
-                Nh = msg.ByteSizeLong();
-                data.resize(Nh);
-                msg.SerializeToArray(data.data(), Nh);
+        for (std::set<std::shared_ptr<webots::Device>>::iterator it = sensors.begin(); it != sensors.end(); ++it) {
+            std::shared_ptr<webots::Accelerometer> accelerometer = std::dynamic_pointer_cast<webots::Accelerometer>(*it);
+            if (accelerometer) {
+                if (time_step % accelerometer->getSamplingPeriod()) {
+                    continue;
+                }
+                AccelerometerMeasurement measurement = *sensorMeasurements->add_accelerometers();
+                measurement.set_name(accelerometer->getName());
+                const double* values = accelerometer->getValues();
+                Vector3 vector3 = *measurement.mutable_value();
+                vector3.set_x(values[0]);
+                vector3.set_y(values[1]);
+                vector3.set_z(values[2]);
+                continue;
+            }
+            std::shared_ptr<webots::Camera> camera = std::dynamic_pointer_cast<webots::Camera>(*it);
+            if (camera) {
+                if (time_step % camera->getSamplingPeriod()) {
+                    continue;
+                }                    
+                CameraMeasurement measurement = *sensorMeasurements->add_cameras();
+                measurement.set_name(camera->getName());
+                measurement.set_width(camera->getWidth());
+                measurement.set_height(camera->getHeight());
+                measurement.set_quality(-1);  // raw image (JPEG compression not yet supported)
+                measurement.set_image((const char *)camera->getImage());
 
-                Nn = htonl(Nh);
+                // testing JPEG compression (impacts the performance)
```

Comment:
Plz no commented code.

Reply:
I guess we can put it back in when they're ready with it.